src/ATen/Scalar.h

```diff
@@ -21,7 +21,7 @@ class AT_API Scalar {
 public:
   Scalar() : Scalar(int64_t(0)) {}
 
-  explicit Scalar(const detail::TensorBase & t)
+  Scalar(const detail::TensorBase & t)
```

Comment:
The explicit prevents horrible overload problems when Scalar and Tensor can appear in the same position in a function list. I believe it only spawns when there is _another_ argument that also needs some implicit conversion.  I am hesitant to remove explicit until we remove all the cases where Scalar and Tensor can appear in the same argument list.

Reply:
I thought about this but figured because nothing got tripped up we were okay; I'll look into this a bit more.

Reply:
Yes I would second not getting rid of explicit constructors. Implicit constructors are really quite annoying.

Reply:
Ok, how about this: we get rid of the implicit conversion and add a Tensor overload for functions like fill, where we want x.fill(y.mean()) to continue working (note: it doesn't work because those functions only have a Scalar argument now).  I don't know if we figure out that set of functions that programmatically or we mark them in the cwrap somehow.  That is against the spirit of https://github.com/zdevito/ATen/issues/117, but we can fix it when we address that issue.

Reply:
I looked through the functions and here are functions where there exists a Scalar argument that doesn't have a Tensor equivalent overload already, grouped by related scalar argument types:

**fills:**
- fill_
- maskedFill_
- indexFill_

**range/min/max:**
- range
- arange
- linspace
- logspace
- histc
- clamp
- clamp_

**exponent/p-norm:**
- norm
- renorm
- renorm_
- dist

**scaling factor/weight:**
- lerp
- lerp_
- add
- add_
- sub
- sub_
- addmm
- addmm_
- addmv
- addmv_
- addr
- addr_
- addbmm
- addbmm_
- baddbmm
- baddbmm_
- baddbmm_
- addcmul
- addcmul_
- addcdiv
- addcdiv_

**tolerance:**
- pstrf

Reply:
I think the most straightforward way of implementing this is just to rely on the handle_zero_dim code; we'll add an option 'only_zero_dim' or similar which means the function should just throw an exception if something that isn't zero_dim is passed in.  Then it's just a matter of choosing which overloads we want to define in Declarations.cwrap.

I'd argue for just the fill functions at this point: I don't really see a use case for the exponent/p-norm, scaling factor/weight, or tolerance being the result of some tensor calculation (at least I haven't used them as such).  I have used them in ranges, but ranges/min/max take multiple values and handle_zero_dim doesn't handle that yet anyway.